octopoes/octopoes/xtdb/query.py

```diff
@@ -0,0 +1,195 @@
+from dataclasses import dataclass, field
+from typing import List, Optional, Type, Union
+
+from octopoes.models import OOI
+from octopoes.models.path import Direction, Path
+from octopoes.models.types import get_abstract_types, get_relations
+
+
+class InvalidField(ValueError):
+    pass
+
+
+class InvalidPath(ValueError):
+    pass
+
+
+@dataclass
+class Query:
+    """Object representing an XTDB query.
+
+        result_type: The OOI Type being queried: executing the query should yield only this OOI Type.
+
+    Example usage:
+
+    >>> query = Query(Network).where(Network, name="test")
+    >>> query = query.where(Finding, ooi=Network)
+    >>> query.format()
+    '
+    {:query {:find [(pull Network [*])] :where [
+        [ Network :Network/name "test" ]
+        [ Finding :Finding/ooi Network ]
+    ]}}
+    '
+    """
+
+    result_type: Type[OOI]
+
+    _where_clauses: List[str] = field(default_factory=list)
+    _find_clauses: List[str] = field(default_factory=list)
+    _limit: Optional[int] = None
+    _offset: Optional[int] = None
+
+    def where(self, ooi_type: Type[OOI], **kwargs) -> "Query":
+        for field_name, value in kwargs.items():
+            self._where_field_is(ooi_type, field_name, value)
+
+        return self
+
+    def format(self) -> str:
+        return "\n" + self._compile(separator="\n    ") + "\n"
+
+    @classmethod
+    def from_path(cls, path: Path) -> "Query":
+        """
+        Create a query from a Path.
+
+        The last segment in the path is assumed to be the queries OOI Type. You can change this by calling the
+        query() method after initialization for the required target OOI Type.
+        """
+
+        ooi_type = path.segments[-1].target_type
+        query = Query(ooi_type)
+
+        for segment in path.segments:
+            if segment.direction is Direction.OUTGOING:
+                query = query.where(segment.source_type, **{segment.property_name: segment.target_type})
+                continue
```

Comment:
I think an `else:` is better here than a `continue` here, because `continue` signals that we skip the rest of the loop for some reason, but that is not really the case.